File: nodejs/base/api/lib/routing/defaults/routes/default-update-builder.ts
import { onResourceUpsert, McmaResource, getTableName } from "@mcma/core";
import { DocumentDatabaseTableProvider } from "@mcma/data";
import { getPublicUrl } from "../../../context-variable-provider-ext";
import { DefaultRouteHandlerConfigurator } from "../default-route-handler-configurator";
import { DefaultRouteBuilder } from "../default-route-builder";

export function defaultUpdateBuilder<T extends McmaResource>(
    dbTableProvider: DocumentDatabaseTableProvider,
    root: string
): DefaultRouteBuilder<T> {
    return new DefaultRouteBuilder<T>(
        "PUT",
        root + "/{id}",
        new DefaultRouteHandlerConfigurator<T>((onStarted, onCompleted) =>
            async (requestContext) => {
                if (onStarted) {
                    const continueRequest = await onStarted(requestContext);
                    if (continueRequest !== undefined && !continueRequest) {
                        return;
                    }
                }

                const resource = requestContext.getRequestBody();
                if (!resource) {
                    requestContext.setResponseBadRequestDueToMissingBody();
                    return;
                }

                onResourceUpsert(resource, getPublicUrl(requestContext) + requestContext.request.path);
                
                const dbTable = await dbTableProvider.get(getTableName(requestContext));
                
                await dbTable.put<T>(requestContext.request.path, resource);

                if (onCompleted) {
                    await onCompleted(requestContext, resource);
                }

                requestContext.setResponseBody(resource);
            }
        )
    );
}

File: dotnet/base/Mcma.Api/Routing/Defaults/Routes/DefaultUpdateBuilder.cs
using System.Net.Http;
using Mcma;
using Mcma.Context;
using Mcma.Data;

namespace Mcma.Api.Routing.Defaults.Routes
{
    public static class DefaultUpdateBuilder
    {
        public static DefaultRouteBuilder<TResource> Get<TResource>(IDocumentDatabaseTableProvider dbTableProvider, string root) where TResource : McmaResource
            =>
            new DefaultRouteBuilder<TResource>(
                HttpMethod.Put,
                root + "/{id}",
                new DefaultRouteHandlerConfigurator<TResource>(
                    (onStarted, onCompleted) =>
                        async requestContext =>
                        {
                            // invoke the start handler, if any
                            if (onStarted != null)
                                if (!await onStarted(requestContext))
                                    return;

                            // ensure the body is set
                            var resource = requestContext.GetRequestBody<TResource>();
                            if (resource == null)
                            {
                                requestContext.SetResponseBadRequestDueToMissingBody();
                                return;
                            }
                            
                            // set properties for upsert
                            resource.OnUpsert(requestContext.CurrentRequestPublicUrl());

                            // upsert the resource
                            var table = await dbTableProvider.GetAsync(requestContext.TableName());
                            await table.PutAsync(requestContext.Request.Path, resource);

                            // invoke the completion handler, if any
                            if (onCompleted != null)
                                await onCompleted(requestContext, resource);

                            // return the new or updated resource as json
                            requestContext.SetResponseBody(resource);
                        }));
    }
}